src/libraries/System.Runtime.Numerics/src/System/Numerics/BigInteger.cs

```diff
@@ -1591,6 +1592,57 @@ public string ToString([StringSyntax(StringSyntaxAttribute.NumericFormat)] strin
             return BigNumber.FormatBigInteger(this, format, NumberFormatInfo.GetInstance(provider));
         }
 
+        private string DebuggerDisplay
+        {
+            get
+            {
+                // For very big numbers, ToString can be too long or even timeout for Visual Studio to display
+                // Display a fast estimated value instead
+
+                // Use ToString for small values
+
+                if (_bits is null || _bits.Length <= 2)
+                {
+                    return ToString();
+                }
+
+                // Estimate the value x as L*2^n, while L is the value of high bits, and n is the length of low bits
+                // Represent L as k*10^i, then x = L*2^n = k*10^(i+n*Log10(2))
+                // Let m=n*Log10(2), the final result would be x = (k*10^(m-[m]))*10^(i+[m])
+
+                ulong highBits = ((ulong)_bits[^1] << kcbitUint) + _bits[^2];
+                int lowBitsCount32 = _bits.Length - 2; // max length of BitInteger is Array.MaxLength/sizeof(uint), which is Array.MaxLength bits
+                double exponentLow = lowBitsCount32 * kcbitUint * Math.Log10(2);
+
+                // Max possible value < 2^Array.MaxLength < 2^int.MaxValue < 10^int.MaxValue
+                // No worrying about int overflow
+                int exponent = (int)exponentLow;
+                double significand = (double)highBits * Math.Pow(10, exponentLow - exponent);
```

Comment:
No. `_bits.Length <= Array.MaxLength / sizeof(uint)` only guarantees the amount of _bytes_ is no more than `Array.MaxLength`, not the amount of _bits_.
```suggestion
                double lowBitsCount32 = _bits.Length - 2; // max length of _bits is Array.MaxLength / sizeof(uint), which unfortunately may have more than int.MaxValue bits.
                double exponentLow = lowBitsCount32 * kcbitUint * Math.Log10(2);

                // Max possible value around 256^Array.MaxLength > 10^int.MaxValue
                // int may overflow.
                long exponent = (long)exponentLow;
                double significand = (double)highBits * Math.Pow(10, exponentLow - exponent);
```

Reply:
Ahh, good point. `sizeof(uint)` is in units of bytes, so it returns `4`, not `32`.

Reply:
I think the issue is that the logic around `MaxLength` matches the initial fix discussed (which is that any `BigInteger` should be able to fit into a `Span<byte>`) but not the secondary fix that was discussed in that we shouldn't have more than `int.MaxValue` bits so that the generic math `GetShortestBitLength` API could do an altogether simpler thing.

As it is currently, the maximum length is `0x1FFF_FFF1` uints, which comes up to `0x7FFF_FFC4` bytes (an approx. `2 GB` allocation) or `0x3_FFFF_FE20` bits.

The requirement that we fit into `int.MaxValue` bits should put the maximum length at `0x3FF_FFFF` uints, which comes up to `0xFFF_FFFC` bytes (an approx. `256 GB` allocation) or `0x7FFF_FFE0` bits.

A `256 GB` allocation for `BigInteger` should be enough for anyone. This is an astronomically big number at `(2^2_147_483_616) - 1` or `~4.10 * 10^646_456_983` and is really pushing the boundaries of what can fit in memory for anything outside supercomputers. Additionally, once you are creating GC allocations this large, you have many other concerns to consider and a different approach using `native memory` is probably better ;)

Reply:
It sounds like during the BigInteger refactor we should come to a consensus on the actual max size (maybe `int.MaxValue` bits) and have every API assume and enforce that cap.